go-controller/pkg/node/helper_linux.go

```diff
@@ -7,35 +7,74 @@ import (
 	"net"
 	"syscall"
 
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
 	"github.com/vishvananda/netlink"
 	kapi "k8s.io/api/core/v1"
+	"k8s.io/klog"
 	utilnet "k8s.io/utils/net"
 )
 
 // getDefaultGatewayInterfaceDetails returns the interface name on
 // which the default gateway (for route to 0.0.0.0) is configured.
-// It also returns the default gateway itself.
-func getDefaultGatewayInterfaceDetails() (string, net.IP, error) {
-	routes, err := netlink.RouteList(nil, syscall.AF_INET)
-	if err != nil {
-		return "", nil, fmt.Errorf("failed to get routing table in node")
+// It also returns the default gateways themselves.
+func getDefaultGatewayInterfaceDetails() (string, []net.IP, error) {
+	var familyList []int
+	var intfName string
+	var gatewayIps []net.IP
+	var needIPv4 bool
+	var needIPv6 bool
+
+	if config.IPv4Mode {
+		familyList = append(familyList, syscall.AF_INET)
+		needIPv4 = true
+	}
+	if config.IPv6Mode {
+		familyList = append(familyList, syscall.AF_INET6)
+		needIPv6 = true
 	}
 
-	for i := range routes {
-		route := routes[i]
-		if route.Dst == nil && route.Gw != nil && route.LinkIndex > 0 {
-			intfLink, err := netlink.LinkByIndex(route.LinkIndex)
-			if err != nil {
-				continue
-			}
-			intfName := intfLink.Attrs().Name
-			if intfName != "" {
-				return intfName, route.Gw, nil
+	for _, family := range familyList {
+		routes, err := netlink.RouteList(nil, family)
```

Comment:
You can use `syscall.AF_UNSPEC` to get back the IPv4 and IPv6 routes together... that might be easier?